---
output: 
  html_document:
    code_download: false
    toc: true                  # table of content true
    toc_depth: 3               # upto three depths of headings (specified by #, ## and ###)
    toc_float: true
    number_sections: true      # if you want number sections at each table header
    theme: united              # many options for theme, this one is my favorite.
    highlight: tango           # specifies the syntax highlighting style
    css: 'style.css'
params:
  gsa_plot:  "gsa.png"
  gsa_result: "gsa.tsv"
title: "`r paste0('Genomics England GWAS-GSA Report' , '') `"
author: ""
date: ""
---

<img src="gel_logo.png" width=100 style="position:absolute;top:4px;bottom:4px;right:4px;" />


```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE)
```

```{r echo = FALSE}
source("DTable.R")

suppressPackageStartupMessages({
library(magrittr)
library(DT)
  
library(htmltools)
library(stringr)
library(dplyr)
library(kableExtra)
   })
```

# About

This report is generated from [GWAS-GSA pipeline](https://github.com/lifebit-ai/gwasgsa/) (Gene-Set Analysis on Genome Wide Association Studies) which uses [MAGMA tool](https://ctg.cncr.nl/software/magma).

The input are GWAS results, such as   
- Summary statistics file  
- PLINK files  

# Visualisations

A dot plots to explore Gene Set Analysis results. Top 10 gene sets used here based on sorting of p-value.

```{r gsa_plot}
figure_number  <- "Figure 1: "
figure_caption <- "gsa plot"
knitr::include_graphics(params$gsa_plot)
htmltools::tags$figcaption( style = 'caption-side: bottom; text-align: center; font-size: 85%%; color: #71879d',
                            htmltools::em(figure_number),
                            htmltools::em(figure_caption))
```

# Results Table

All the results generated form MAGMA tool with an additional column of gene names. This table is sorted based on p-value.

<br><br>

```{r gsa_result}
table   <- read.table(params$gsa_result, header = T, sep = "\t")
DTable(table)
```

